src/libraries/System.Formats.Tar/tests/TarEntry/GnuTarEntry.Tests.cs

```diff
@@ -93,5 +94,469 @@ public void SupportedEntryType_Fifo()
             SetFifo(fifo);
             VerifyFifo(fifo);
         }
+
+        [Theory]
+        [InlineData(false)]
+        [InlineData(true)]
+        public void DataOffset_RegularFile(bool canSeek)
+        {
+            byte expectedData = 5;
+            using MemoryStream ms = new();
+            using (TarWriter writer = new(ms, leaveOpen: true))
+            {
+                GnuTarEntry entry = new GnuTarEntry(TarEntryType.RegularFile, InitialEntryName);
+                entry.DataStream = new MemoryStream();
+                entry.DataStream.WriteByte(5);
+                entry.DataStream.Position = 0;
+                writer.WriteEntry(entry);
+            }
+            ms.Position = 0;
+
+            using Stream streamToRead = new WrappedStream(ms, canWrite: true, canRead: true, canSeek: canSeek);
+            using TarReader reader = new(streamToRead);
+            TarEntry actualEntry = reader.GetNextEntry();
+            Assert.NotNull(actualEntry);
+
+            // Then it writes the actual regular file entry, containing:
+            // * 512 bytes of the regular tar header
+            // Totalling 2560.
+            // The regular file data section starts on the next byte.
+            long expectedDataOffset = canSeek ? 512 : -1;
+            Assert.Equal(expectedDataOffset, actualEntry.DataOffset);
```

Comment:
```suggestion
            Assert.Equal(expectedDataOffset, actualEntry.DataOffset);
            Assert.Equal(expectedData, (byte)actualEntry.DataStream.ReadByte());
```

Might be good to assert that the TarEntry reads the same data that the stream at that position does.

Reply:
Extended some tests to confirm this https://github.com/dotnet/runtime/pull/105007/commits/3e4fd4c9145732c7798b0a1549fc7efb6a72ccb0